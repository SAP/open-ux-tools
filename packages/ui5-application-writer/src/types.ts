export interface Package {
    name: string;
    version?: string;
    description?: string;
    scripts?: Record<string, string>;
    dependencies?: Record<string, string>;
    devDependencies?: Record<string, string>;
    ui5?: {
        dependencies?: string[];
    };
}

export interface App {
    id: string;
    version?: string;
    title?: string;
    description?: string;
    baseComponent?: string; // UI5 Component `required` by Component.js
    startFile?: string;
    localStartFile?: string;

    /**
     * Defines the template used by the generation tool
     */
    sourceTemplate?: {
        /**
         * Identifies the source template used to generate the application. Set to empty string if not provided.
         */
        id?: string;
        /**
         * Source template version. Set to empty string if not provided.
         */
        version?: string;
        /**
         * Source template tools id. Optional property. Will not be written to manifest.json if not provided
         */
        toolsId?: string;
    };
}

export type UI5Framework = 'SAPUI5' | 'OpenUI5';

export interface UI5 {
    framework: UI5Framework;
    frameworkUrl: string;
    minUI5Version: string;
    version: string;
    localVersion: string;
    typesVersion: string;
    descriptorVersion: string;
    ui5Libs: string | string[];
    ui5Theme: string;
}

// Additional configurable features
export interface AppOptions {
    codeAssist: boolean; // Enables code assist
    eslint: boolean; // Enables Fiori linting
    sapux: boolean; // Enables Fiori Tools Suite integration

    /**
     * Enables loading of re-use libs
     */
    loadReuseLibs: boolean;

    /**
     * Enable Typescript support and generate all files in Typescript instead of Javascript
     * Enabling Typescript will disable JavaScript codeAssist if also enabled. Typescript code assist will be used instead
     */
    typescript: boolean;

    /**
     * Enable support for importing functionality from npm packages in UI5 applications
     */
    npmPackageConsumption: boolean;
}

export interface Ui5App {
    app: App;
    appOptions?: Partial<AppOptions>;
    ui5?: Partial<UI5>;
    package: Package;
}
