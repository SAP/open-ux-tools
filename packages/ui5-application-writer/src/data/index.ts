import type { App, UI5, AppOptions, Package, Ui5App } from '../types';
import { mergeApp, packageDefaults, mergeUi5, mergeObjects, getSpecTagVersion } from './defaults';
import { validate } from './validators';

/**
 * Merges Ui5App instance with default properties.
 * Replaces undefined mandatory properties with default values.
 *
 * @param {Ui5App} ui5App - the Ui5App instance
 * @returns {Ui5App} - a new Ui5App instance with all required defaults set
 */
export function mergeWithDefaults(ui5App: Ui5App): {
    app: App;
    appOptions: Partial<AppOptions>;
    ui5: UI5;
    package: Package;
} {
    validate(ui5App);
    ui5App.app = mergeApp(ui5App.app);
    ui5App.appOptions = ui5App.appOptions || {};
    // if typescript and codeAssist is enabled disable codeAssist
    if (ui5App.appOptions.typescript && ui5App.appOptions.codeAssist) {
        ui5App.appOptions.codeAssist = false;
    }
    ui5App.ui5 = mergeUi5(ui5App.ui5 || {}, ui5App.appOptions);
    ui5App.package = mergeObjects(packageDefaults(ui5App.package.version, ui5App.app.description), ui5App.package);

    if (ui5App.appOptions.sapux) {
        ui5App.package.devDependencies = ui5App.package.devDependencies || {};
        ui5App.package.devDependencies['@sap/ux-specification'] = getSpecTagVersion(ui5App.ui5.version);
    }

    return ui5App as {
        app: App;
        appOptions: Partial<AppOptions>;
        ui5: UI5;
        package: Package;
    };
}
