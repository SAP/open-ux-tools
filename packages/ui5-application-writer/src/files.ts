import { readdirSync, statSync } from 'fs';
import { join } from 'path';

/**
 * @description Returns a flat list of all file paths under a directory tree,
 * recursing through all subdirectories
 * @param {string} dir - the directory to walk
 * @returns {string[]} - array of file path strings
 * @throws if an error occurs reading a file path
 */
export function getFilePaths(dir: string): string[] | [] {
    return readdirSync(dir).reduce((files: string[], entry: string) => {
        const entryPath = join(dir, entry);
        const isDirectory = statSync(entryPath).isDirectory();
        return isDirectory ? [...files, ...getFilePaths(entryPath)] : [...files, entryPath];
    }, []);
}
