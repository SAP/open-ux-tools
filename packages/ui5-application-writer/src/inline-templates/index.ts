import { Minimatch } from 'minimatch';
import coreGitignore from '../../templates/core/gitignore.ejs';
import corePackageJson from '../../templates/core/package.json.ejs';
import coreUi5LocalYaml from '../../templates/core/ui5-local.yaml.ejs';
import coreUi5Yaml from '../../templates/core/ui5.yaml.ejs';
import coreWebappComponent from '../../templates/core/webapp/Component.js.ejs';
import coreWebappIndex from '../../templates/core/webapp/index.html.ejs';
import coreWebappManifest from '../../templates/core/webapp/manifest.json.ejs';
import coreWebappI18nProperties from '../../templates/core/webapp/i18n/i18n.properties.ejs';

import optionalCodeAssistEslintRc from '../../templates/optional/codeAssist/.eslintrc.ejs';
import optionalCodeAssistPackageJson from '../../templates/optional/codeAssist/package.json.ejs';
import optionalCodeAssistTsconfig from '../../templates/optional/codeAssist/tsconfig.json.ejs';

import optionalEslintEslintRc from '../../templates/optional/eslint/.eslintrc.ejs';
import optionalEslintPackageJson from '../../templates/optional/eslint/package.json.ejs';

import loadReuseLibs1120WebappTest from '../../templates/optional/loadReuseLibs/1.120.0/webapp/test/locate-reuse-libs.js.ejs';
import loadReuseLibs171WebappTest from '../../templates/optional/loadReuseLibs/1.71.0/webapp/test/locate-reuse-libs.js.ejs';

import optionalNpmPackageConsumptionPackageJson from '../../templates/optional/npmPackageConsumption/package.json.ejs';
import optionalSapuxPackageJson from '../../templates/optional/sapux/package.json.ejs';

import optionalTypescriptWebappComponent from '../../templates/optional/typescript/webapp/Component.ts.ejs';
import optionalTypescriptEslintRc from '../../templates/optional/typescript/.eslintrc.ejs';
import optionalTypescriptPackageJson from '../../templates/optional/typescript/package.json.ejs';
import optionalTypescriptTsconfig from '../../templates/optional/typescript/tsconfig.json.ejs';

const templates = {
    'core/.gitignore': coreGitignore,
    'core/package.json': corePackageJson,
    'core/ui5-local.yaml': coreUi5LocalYaml,
    'core/ui5.yaml': coreUi5Yaml,
    'core/webapp/Component.js': coreWebappComponent,
    'core/webapp/index.html': coreWebappIndex,
    'core/webapp/manifest.json': coreWebappManifest,
    'core/webapp/i18n/i18n.properties': coreWebappI18nProperties,
    'optional/codeAssist/.eslintrc': optionalCodeAssistEslintRc,
    'optional/codeAssist/package.json': optionalCodeAssistPackageJson,
    'optional/codeAssist/tsconfig.json': optionalCodeAssistTsconfig,
    'optional/eslint/.eslintrc': optionalEslintEslintRc,
    'optional/eslint/package.json': optionalEslintPackageJson,
    'optional/npmPackageConsumption/package.json': optionalNpmPackageConsumptionPackageJson,
    'optional/loadReuseLibs/1.120.0/webapp/test/locate-reuse-libs.js': loadReuseLibs1120WebappTest,
    'optional/loadReuseLibs/1.71.0/webapp/test/locate-reuse-libs.js': loadReuseLibs171WebappTest,
    'optional/sapux/package.json': optionalSapuxPackageJson,
    'optional/typescript/webapp/Component.ts': optionalTypescriptWebappComponent,
    'optional/typescript/.eslintrc': optionalTypescriptEslintRc,
    'optional/typescript/package.json': optionalTypescriptPackageJson,
    'optional/typescript/tsconfig.json': optionalTypescriptTsconfig
};

/**
 * Recursively retrieves all files and their content from the templates JSON object.
 *
 * @param {string} basePath - The base path in the templates object to start from (e.g., 'core' or 'optional/eslint').
 * @param {Function} [processDestinationPath] - Optional callback to process destination paths.
 * @param {string[]} [ignore] - Optional array of glob file paths to ignore.
 * @returns {Map<string, string>} - A map where keys are file paths and values are their content.
 */
export function getFilesFromTemplates(
    basePath: string,
    processDestinationPath?: (path: string) => string,
    ignore?: string[]
): Map<string, string> {
    const result = new Map<string, string>();

    for (const [key, value] of Object.entries(templates)) {
        if (key.startsWith(basePath)) {
            if (ignore?.some((pattern) => new Minimatch(pattern).match(key))) {
                continue;
            }
            let destinationPath = key;
            // Remove the basePath from the destination path
            destinationPath = destinationPath.replace(`${basePath}/`, '');
            if (processDestinationPath) {
                destinationPath = processDestinationPath(destinationPath);
            }
            result.set(destinationPath, value);
        }
    }

    return result;
}
