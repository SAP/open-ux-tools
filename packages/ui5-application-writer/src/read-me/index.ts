import { ReadMe } from './types';
import type { TemplateWriter, ApplyTemplateFunction } from './types';
import { t } from '../i18n';
import { join } from 'path';
import { Editor } from 'mem-fs-editor';

/**
 * Merges the provided ReadMe configuration with default values to be used by read me templates.
 * 
 * @param {Partial<ReadMe>} readMeConfig - The partial configuration object to merge with defaults.
 * @returns {ReadMe} The merged configuration object with defaults applied.
 */
function mergeWithDefaults(readMeConfig: Partial<ReadMe>): ReadMe {
    const defaults: ReadMe = {
        genDate: new Date().toString(),
        genPlatform: '',
        dataSourceLabel: '',
        metadataFilename: '',
        serviceUrl: 'N/A',
        projectName: '',
        projectTitle: '',
        projectDescription: '',
        projectNamespace: '',
        ui5Theme: '',
        projectUI5Version: '',
        enableCodeAssist: false,
        enableEslint: false,
        enableTypeScript: false,
        showMockDataInfo: false,
        genVersion: '',
        templateLabel: '',
        genId: '',
        additionalEntries: [],
        launchText: t('TEXT_LAUNCH_DEFAULT')
    };

    return { ...defaults, ...readMeConfig };
}

/**
 * Returns a function that writes a template to the file system.
 * @param {object} templateWriter - The template writer object.
 * @param {string} templateWriter.fileName - The name of the file to write.
 * @param {string} templateWriter.destPath - The destination path to write the file to.
 * @param {Editor} templateWriter.fsEditor - The file system editor.
 * @returns The function that writes the template to the file system.
 */
function getTemplateWriter({ fileName, destPath, fsEditor}: TemplateWriter): ApplyTemplateFunction {
    // Determine if the template path is bundled (prod) or not (test)
    const isBundledPath = fsEditor.exists(join(__dirname, '..', 'templates', fileName));
    return <P>(path: string, properties: P): void => {
        // Compute the template source path based on whether it's bundled or not
        const templateSourcePath = join(__dirname, isBundledPath ? '' : '..', '..', 'templates', path);
        // Compute the destination path, removing specific substrings from the path
        const templateDestPath = join(destPath, path.replace(/\.tmpl|-cap-tmpl/g, ''));
        // Copy the template with the provided properties
        fsEditor.copyTpl(templateSourcePath, templateDestPath, properties!);
    };
}

/**
 * Generates a README file at the specified destination path using the provided configuration and file system editor.
 *
 * @param {string} destPath - The desitination path where the README file will be created.
 * @param {Partial<ReadMe>} readMeConfig - The configuration object containing the details to be included in the README file. Properties in this object are optional.
 * @param {Editor} fs - The file system editor instance used to write the README file.
 * @returns {Editor} The file system editor instance used to write the README file.
 *
 * @example
 * const readMeConfig = {
 *     projectName: "MyProject",
 *     projectTitle: "My Project Title",
 *     projectNamespace: "com.example",
 *     projectDescription: "A description of my project.",
 *     ui5Theme: "sap_belize",
 *     projectUI5Version: "1.84.0",
 *     enableCodeAssist: true,
 *     enableEslint: true,
 *     enableTypeScript: false,
 *     genId: "generator-id",
 *     genVersion: "1.0.0",
 *     templateLabel: "List Report Page",
 *     genDate: "2023-06-01",
 *     genPlatform: "Web",
 *     dataSourceLabel: "OData V4",
 *     serviceUrl: "https://example.com/service",
 *     showMockDataInfo: true,
 *     additionalEntries: [
 *         { label: "Custom Entry", value: "Custom Value" }
 *     ],
 *     launchText: "To launch the project, run `npm start`."
 * };
 *
 * generateReadMe("./README.md", readMeConfig, editor);
 */
export function generateReadMe(destPath: string, readMeConfig: Partial<ReadMe>, fs: Editor): Editor {
    const config: ReadMe = mergeWithDefaults(readMeConfig);
    // Apply the configuration to generate the README.md file
    getTemplateWriter({ fileName: 'README.md', destPath, fsEditor: fs })<ReadMe>('README.md', config);
    return fs;
}

export { ReadMe };
