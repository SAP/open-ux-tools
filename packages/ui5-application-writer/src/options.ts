import { join } from 'path';
import type { Editor } from 'mem-fs-editor';
import { render } from 'ejs';
import type { Ui5App } from './types';
import { getFilePaths } from './files';
import type { UI5Config } from '@sap-ux/ui5-config';
import { ui5NPMSupport, ui5TSSupport } from './data/ui5Libs';
import { mergeObjects, UI5_DEFAULT } from './data/defaults';

/**
 * Input required to enable optional features.
 */
export interface FeatureInput {
    ui5App: { app: { id: string; baseComponent?: string } };
    fs: Editor;
    basePath: string;
    tmplPath: string;
    ui5Configs: UI5Config[];
}

/**
 * Copy all template files into the target project.
 *
 * @param name the name of the optional feature
 * @param input collection of input properties
 * @param input.ui5App ui5 app config
 * @param input.fs reference to the mem-fs instance
 * @param input.basePath project base path
 * @param input.tmplPath template basepath
 */
function copyTemplates(name: string, { ui5App, fs, basePath, tmplPath }: FeatureInput) {
    const optTmplDirPath = join(tmplPath, 'optional', `${name}`);
    const optTmplFilePaths = getFilePaths(optTmplDirPath);
    optTmplFilePaths.forEach((optTmplFilePath) => {
        const relPath = optTmplFilePath.replace(optTmplDirPath, '');
        const outPath = join(basePath, relPath);
        // Extend or add
        if (!fs.exists(outPath)) {
            fs.copyTpl(optTmplFilePath, outPath, ui5App, undefined, {
                globOptions: { dot: true }
            });
        } else {
            const add = JSON.parse(render(fs.read(optTmplFilePath), ui5App, {}));
            const existingFile = JSON.parse(fs.read(outPath));
            const merged = mergeObjects(existingFile, add);
            fs.writeJSON(outPath, merged);
        }
    });
}

/**
 * Factory functions for applying optional features.
 */
const factories: { [key: string]: (input: FeatureInput) => void } = {
    codeAssist: (input: FeatureInput) => copyTemplates('codeAssist', input),
    eslint: (input: FeatureInput) => copyTemplates('eslint', input),
    loadReuseLibs: (input: FeatureInput) => copyTemplates('loadReuseLibs', input),
    sapux: (input: FeatureInput) => copyTemplates('sapux', input),
    typescript: enableTypescript,
    npmPackageConsumption: enableNpmPackageConsumption
};

/**
 * Enable typescript for the given input.
 *
 * @param input Input required to enable the optional typescript features
 * @param keepOldComponent if set to true then the old Component.js will be renamed but kept.
 */
export function enableTypescript(input: FeatureInput, keepOldComponent: boolean = false) {
    input.ui5App.app.baseComponent = input.ui5App.app.baseComponent ?? UI5_DEFAULT.BASE_COMPONENT;
    copyTemplates('typescript', input);
    input.ui5Configs.forEach((ui5Config) => {
        ui5Config.addCustomMiddleware([ui5TSSupport.middleware]);
        ui5Config.addCustomTasks([ui5TSSupport.task]);
    });
    const compPath = join(input.basePath, 'webapp/Component.js');
    if (keepOldComponent) {
        input.fs.move(compPath, `${compPath}.old`);
    } else {
        input.fs.delete(compPath);
    }
}

/**
 * Enable npm module import for the given input.
 *
 * @param input Input required to enable the optional npm modules import
 */
export function enableNpmPackageConsumption(input: FeatureInput) {
    copyTemplates('npmPackageConsumption', input);
    input.ui5Configs.forEach((ui5Config) => {
        ui5Config.addCustomMiddleware([ui5NPMSupport.middleware]);
        ui5Config.addCustomTasks([ui5NPMSupport.task]);
    });
}

/**
 * Check if the ui5 app config requires optional features to be enabled and if yes, enable them.
 *
 * @param ui5App ui5 app config
 * @param fs reference to the mem-fs instance
 * @param basePath project base path
 * @param tmplPath template basepath
 * @param ui5Configs available UI5 configs
 */
export function applyOptionalFeatures(
    ui5App: Ui5App,
    fs: Editor,
    basePath: string,
    tmplPath: string,
    ui5Configs: UI5Config[]
) {
    if (ui5App.appOptions) {
        Object.entries(ui5App.appOptions).forEach(([key, value]) => {
            if (value === true) {
                factories[key]?.({ ui5App, fs, basePath, tmplPath, ui5Configs });
            }
        });
    }
}
