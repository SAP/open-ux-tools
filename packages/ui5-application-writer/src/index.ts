import { join } from 'path';
import { create as createStorage } from 'mem-fs';
import type { Editor } from 'mem-fs-editor';
import { create } from 'mem-fs-editor';
import type { App, AppOptions, Package, UI5 } from './types';
import { UI5Config } from '@sap-ux/ui5-config';
import type { Manifest } from '@sap-ux/project-access';
import { mergeWithDefaults } from './data';
import { ui5TSSupport } from './data/ui5Libs';
import { applyOptionalFeatures, enableTypescript as enableTypescriptOption } from './options';
import { Ui5App } from './types';
import { getEsmTypesVersion } from './data/defaults';

/**
 * Writes the template to the memfs editor instance.
 *
 * @param basePath - the base path
 * @param ui5AppConfig - the Ui5App instance
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
async function generate(basePath: string, ui5AppConfig: Ui5App, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    const ui5App: { app: App; appOptions: Partial<AppOptions>; ui5: UI5; package: Package } =
        mergeWithDefaults(ui5AppConfig);

    const tmplPath = join(__dirname, '..', 'templates');

    const ignore = [ui5AppConfig.appOptions?.typescript ? '**/*.js' : '**/*.ts'];
    fs.copyTpl(join(tmplPath, 'core', '**/*.*'), join(basePath), ui5App, undefined, {
        globOptions: { dot: true, ignore },
        processDestinationPath: (filePath: string) => filePath.replace(/gitignore.tmpl/g, '.gitignore')
    });

    // ui5.yaml
    const ui5ConfigPath = join(basePath, 'ui5.yaml');
    const ui5Config = await UI5Config.newInstance(fs.read(ui5ConfigPath));
    ui5Config.addFioriToolsProxydMiddleware({
        ui5: {
            url: ui5App.ui5?.frameworkUrl
        }
    });
    ui5Config.addFioriToolsAppReloadMiddleware();

    // ui5-local.yaml
    const ui5LocalConfigPath = join(basePath, 'ui5-local.yaml');
    const ui5LocalConfig = await UI5Config.newInstance(fs.read(ui5LocalConfigPath));
    ui5LocalConfig.addUI5Framework(
        ui5App.ui5.framework,
        ui5App.ui5.localVersion,
        ui5App.ui5.ui5Libs as string[],
        ui5App.ui5.ui5Theme
    );
    ui5LocalConfig.addFioriToolsAppReloadMiddleware();

    // Add optional features
    applyOptionalFeatures(ui5App, fs, basePath, tmplPath, [ui5Config, ui5LocalConfig]);

    // write ui5 yamls
    fs.write(ui5ConfigPath, ui5Config.toString());
    fs.write(ui5LocalConfigPath, ui5LocalConfig.toString());

    return fs;
}

/**
 * Check if Typescript is enabled for the given path.
 *
 * @param basePath - the base path
 * @param fs - the memfs editor instance
 * @returns true is the project is ready for typescript code
 */
async function isTypescriptEnabled(basePath: string, fs?: Editor): Promise<boolean> {
    if (!fs) {
        fs = create(createStorage());
    }

    // check middlewares and tasks
    const ui5Config = await UI5Config.newInstance(fs.read(join(basePath, 'ui5.yaml')));
    if (!ui5Config.findCustomMiddleware(ui5TSSupport.middleware.name)) {
        return false;
    }
    if (!ui5Config.findCustomTask(ui5TSSupport.task.name)) {
        return false;
    }

    return true;
}

/**
 * Enable typescript in an existing valid UI5 project.
 *
 * @param basePath - the base path
 * @param fs - the memfs editor instance
 * @returns the updated memfs editor instance
 */
async function enableTypescript(basePath: string, fs?: Editor): Promise<Editor> {
    if (!fs) {
        fs = create(createStorage());
    }
    const manifestPath = join(basePath, 'webapp/manifest.json');
    const ui5ConfigPath = join(basePath, 'ui5.yaml');
    if (!fs.exists(manifestPath)) {
        throw new Error(`Invalid project folder. Cannot find required file ${manifestPath}`);
    }
    if (!fs.exists(ui5ConfigPath)) {
        throw new Error(`Invalid project folder. Cannot find required file ${ui5ConfigPath}`);
    }
    const manifest = fs.readJSON(manifestPath) as any as Manifest;
    const ui5Config = await UI5Config.newInstance(fs.read(ui5ConfigPath));

    const tmplPath = join(__dirname, '..', 'templates');
    const ui5App = {
        app: manifest['sap.app'],
        ui5: {
            typesVersion: getEsmTypesVersion(manifest['sap.ui5']?.dependencies?.minUI5Version)
        }
    };
    enableTypescriptOption({ basePath, fs, ui5Configs: [ui5Config], tmplPath, ui5App }, true);

    fs.write(ui5ConfigPath, ui5Config.toString());

    return fs;
}

export { Ui5App, generate, enableTypescript, isTypescriptEnabled };
export { App, Package, UI5, AppOptions };
