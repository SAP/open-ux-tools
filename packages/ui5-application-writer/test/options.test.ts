import { generate } from '../src';
import { join } from 'path';
import { removeSync } from 'fs-extra';

describe('UI5 templates', () => {
    const debug = !!process.env['UX_DEBUG'];
    const outputDir = join(__dirname, '/test-output');

    const baseAppConfig = {
        app: {
            id: 'app.with.namespace',
            title: 'Test App Title',
            description: 'Test App Description'
        },
        'package': {
            name: 'testPackageName'
        }
    };

    beforeAll(() => {
        removeSync(outputDir); // even for in memory
    });

    it('generates options: `codeAssist, eslint, sapux`', async () => {
        const projectDir = join(outputDir, 'testapp_options');
        const fs = await generate(projectDir, {
            ...baseAppConfig,
            appOptions: {
                codeAssist: true,
                eslint: true,
                sapux: true
            }
        });
        expect(fs.dump(projectDir)).toMatchSnapshot();
        return new Promise((resolve) => {
            // write out the files for debugging
            if (debug) {
                fs.commit(resolve);
            } else {
                resolve(true);
            }
        });
    });

    it('option: `loadReuseLibs`', async () => {
        const projectDir = join(outputDir, 'testapp_loadReuseLibs');
        const fs = await generate(projectDir, {
            ...baseAppConfig,
            appOptions: {
                loadReuseLibs: true
            }
        });
        expect(fs.dump(projectDir)).toMatchSnapshot();
        return new Promise((resolve) => {
            // write out the files for debugging
            if (debug) {
                fs.commit(resolve);
            } else {
                resolve(true);
            }
        });
    });

    it('generates options: `sapux` with specific version', async () => {
        const projectDir = join(outputDir, 'testapp_options');
        const fs = await generate(projectDir, {
            ...baseAppConfig,
            appOptions: {
                sapux: true
            },
            ui5: {
                version: '1.92.1'
            }
        });
        expect(fs.dump(projectDir)).toMatchSnapshot();
        return new Promise((resolve) => {
            // write out the files for debugging
            if (debug) {
                fs.commit(resolve);
            } else {
                resolve(true);
            }
        });
    });

    it('option: `typescript, npm modules and Fiori tools`', async () => {
        const projectDir = join(outputDir, 'testapp_typescript');
        const fs = await generate(
            projectDir,
            Object.assign(baseAppConfig, {
                appOptions: {
                    typescript: true,
                    sapux: true,
                    npmPackageConsumption: true
                }
            })
        );
        expect(fs.dump(projectDir)).toMatchSnapshot();
        return new Promise((resolve) => {
            // write out the files for debugging
            if (debug) {
                fs.commit(resolve);
            } else {
                resolve(true);
            }
        });
    });

    it('option: `typescript and code assist` to check for conflicts', async () => {
        const projectDir = join(outputDir, 'testapp_tsandcodeassist');
        const fs = await generate(
            projectDir,
            Object.assign(baseAppConfig, {
                appOptions: {
                    typescript: true,
                    codeAssist: true
                }
            })
        );
        expect(fs.read(join(projectDir, 'package.json'))).toMatchSnapshot();
        expect(fs.read(join(projectDir, 'tsconfig.json'))).toMatchSnapshot();
        expect(fs.read(join(projectDir, '.eslintrc'))).toMatchSnapshot();
        return new Promise((resolve) => {
            // write out the files for debugging
            if (debug) {
                fs.commit(resolve);
            } else {
                resolve(true);
            }
        });
    });
});
