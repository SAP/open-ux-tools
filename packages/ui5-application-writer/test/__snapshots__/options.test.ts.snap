// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UI5 templates generates options 1`] = `
Object {
  ".eslintrc": Object {
    "contents": "{
  \\"plugins\\": [
    \\"@sap/ui5-jsdocs\\",
    \\"eslint-plugin-fiori-custom\\"
  ],
  \\"extends\\": [
    \\"plugin:@sap/ui5-jsdocs/recommended\\",
    \\"eslint:recommended\\",
    \\"./node_modules/eslint-plugin-fiori-custom/configure.eslintrc\\"
  ],
  \\"ignorePatterns\\": [
    \\"webapp/localservice/**\\",
    \\"webapp/test/**\\",
    \\"dist/**\\"
  ]
}
",
    "state": "modified",
  },
  ".gitignore": Object {
    "contents": "node_modules/
dist/
.scp/
.env
Makefile*.mta
mta_archives
mta-*
resources
archive.zip
.*_mta_build_tmp",
    "state": "modified",
  },
  "package.json": Object {
    "contents": "{
  \\"name\\": \\"testpackagename\\",
  \\"version\\": \\"0.0.1\\",
  \\"private\\": true,
  \\"description\\": \\"Test App Description\\",
  \\"keywords\\": [
    \\"ui5\\",
    \\"openui5\\",
    \\"sapui5\\"
  ],
  \\"main\\": \\"webapp/index.html\\",
  \\"dependencies\\": {},
  \\"devDependencies\\": {
    \\"@ui5/cli\\": \\"^2.14.1\\",
    \\"@sap/ux-ui5-tooling\\": \\"1\\",
    \\"eslint\\": \\"7.32.0\\",
    \\"@sap/eslint-plugin-ui5-jsdocs\\": \\"2.0.5\\",
    \\"@sapui5/ts-types\\": \\"1.92.2\\",
    \\"eslint-plugin-fiori-custom\\": \\"2.2.1\\",
    \\"@babel/eslint-parser\\": \\"7.14.7\\"
  },
  \\"scripts\\": {
    \\"start\\": \\"ui5 serve --config=ui5.yaml --open index.html\\",
    \\"start-local\\": \\"ui5 serve --config=ui5-local.yaml --open index.html\\",
    \\"build\\": \\"ui5 build --config=ui5.yaml --clean-dest --dest dist\\",
    \\"lint\\": \\"eslint ./ --rulesdir ./node_modules/eslint-plugin-fiori-custom/lib/rules/\\"
  },
  \\"ui5\\": {
    \\"dependencies\\": [
      \\"@sap/ux-ui5-tooling\\"
    ]
  },
  \\"sapux\\": true
}
",
    "state": "modified",
  },
  "tsconfig.json": Object {
    "contents": "{
  \\"compilerOptions\\": {
    \\"module\\": \\"none\\",
    \\"noEmit\\": true,
    \\"checkJs\\": true,
    \\"allowJs\\": true,
    \\"types\\": [
      \\"@sapui5/ts-types\\"
    ]
  }
}",
    "state": "modified",
  },
  "ui5-local.yaml": Object {
    "contents": "# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: \\"2.5\\"
metadata:
  name: testAppId
type: application
framework:
  name: SAPUI5
  version: 1.95.0
  libraries:
    - name: sap.m
    - name: sap.ui.core
    - name: themelib_sap_fiori_3
server:
  customMiddleware:
    - name: fiori-tools-appreload
      afterMiddleware: compression
      configuration:
        port: 35729
        path: webapp
        delay: 300
",
    "state": "modified",
  },
  "ui5.yaml": Object {
    "contents": "# yaml-language-server: $schema=https://sap.github.io/ui5-tooling/schema/ui5.yaml.json

specVersion: \\"2.5\\"
metadata:
  name: testAppId
type: application
server:
  customMiddleware:
    - name: fiori-tools-proxy
      afterMiddleware: compression
      configuration:
        ignoreCertError: false # If set to true, certificate errors will be ignored. E.g. self-signed certificates will be accepted
        ui5:
          path:
            - /resources
            - /test-resources
          url: https://ui5.sap.com
          version: '' # The UI5 version, for instance, 1.78.1. Empty string means latest version
    - name: fiori-tools-appreload
      afterMiddleware: compression
      configuration:
        port: 35729
        path: webapp
        delay: 300
",
    "state": "modified",
  },
  "webapp/Component.js": Object {
    "contents": "sap.ui.define(
    [\\"sap/ui/core/UIComponent\\"],
    function (UIComponent) {
        \\"use strict\\";

        return UIComponent.extend(\\"testAppId.Component\\", {
            metadata: {
                manifest: \\"json\\"
            },

            /**
             * The component is initialized by UI5 automatically during the startup of the app and calls the init method once.
             * @public
             * @override
             */
            init: function () {
                // call the base component's init function
                UIComponent.prototype.init.apply(this, arguments);

                // enable routing
                this.getRouter().initialize();
            }
        });
    }
);",
    "state": "modified",
  },
  "webapp/i18n/i18n.properties": Object {
    "contents": "# This is the resource bundle for testAppId

#Texts for manifest.json

#XTIT: Application name
appTitle=Test App Title

#YDES: Application description
appDescription=Test App Description",
    "state": "modified",
  },
  "webapp/index.html": Object {
    "contents": "<!DOCTYPE html>
<html>
<head>
    <meta charset=\\"UTF-8\\">
    <meta name=\\"viewport\\" content=\\"width=device-width, initial-scale=1.0\\">
    <meta http-equiv=\\"X-UA-Compatible\\" content=\\"IE=edge\\">
    <title>Test App Title</title>
    <style>
        html, body, body > div, #container, #container-uiarea {
            height: 100%;
        }
    </style>
    <script
        id=\\"sap-ui-bootstrap\\"
        src=\\"resources/sap-ui-core.js\\"
        data-sap-ui-theme=\\"sap_fiori_3\\"
        data-sap-ui-resourceroots='{
            \\"testAppId\\": \\"./\\"
        }'
        data-sap-ui-oninit=\\"module:sap/ui/core/ComponentSupport\\"
        data-sap-ui-compatVersion=\\"edge\\"
        data-sap-ui-async=\\"true\\"
        data-sap-ui-frameOptions=\\"trusted\\"
    ></script>
</head>
<body class=\\"sapUiBody sapUiSizeCompact\\" id=\\"content\\">
    <div
        data-sap-ui-component
        data-name=\\"testAppId\\"
        data-id=\\"container\\"
        data-settings='{\\"id\\" : \\"testAppId\\"}'
        data-handle-validation=\\"true\\"
    ></div>
</body>
</html>",
    "state": "modified",
  },
  "webapp/manifest.json": Object {
    "contents": "{
    \\"_version\\": \\"1.12.0\\",
    \\"sap.app\\": {
        \\"id\\": \\"testAppId\\",
      \\"type\\": \\"application\\",
      \\"i18n\\": \\"i18n/i18n.properties\\",
      \\"applicationVersion\\": {
        \\"version\\": \\"0.0.1\\"
      },
      \\"title\\": \\"{{appTitle}}\\",
      \\"description\\": \\"{{appDescription}}\\",
      \\"resources\\": \\"resources.json\\"
    },
    \\"sap.ui\\": {
      \\"technology\\": \\"UI5\\",
      \\"icons\\": {
        \\"icon\\": \\"\\",
        \\"favIcon\\": \\"\\",
        \\"phone\\": \\"\\",
        \\"phone@2\\": \\"\\",
        \\"tablet\\": \\"\\",
        \\"tablet@2\\": \\"\\"
      },
      \\"deviceTypes\\": {
        \\"desktop\\": true,
        \\"tablet\\": true,
        \\"phone\\": true
      }
    },
    \\"sap.ui5\\": {
      \\"flexEnabled\\": true,
      \\"dependencies\\": {
        \\"minUI5Version\\": \\"1.60\\",
        \\"libs\\": {
          \\"sap.m\\": {},
          \\"sap.ui.core\\": {}
        }
      },
      \\"contentDensities\\": {
        \\"compact\\": true,
        \\"cozy\\": true
      },
      \\"models\\": {
        \\"i18n\\": {
          \\"type\\": \\"sap.ui.model.resource.ResourceModel\\",
          \\"settings\\": {
            \\"bundleName\\": \\"testAppId.i18n.i18n\\"
          }
        }
      },
      \\"resources\\": {
        \\"css\\": []
      },
      \\"routing\\": {
        \\"config\\": {},
        \\"routes\\": [],
        \\"targets\\": {}
      }
    }
  }
",
    "state": "modified",
  },
}
`;
